content/docs/scripting.org

```diff
@@ -0,0 +1,180 @@
+
+In order to do that, we have to interact with processes and manipulate string, we are going to see step by step how to do that in our "script":
+
+** Implementation and comparaison with an other scripting language
+I decided to write the same script, first in Zig and then in Python, in order to compare the two languages and see if Zig is a good choice for scripting.
+
+We are going to go through the different steps of the script and see how it is done in Zig and in Python, compare the 2 and see which is best suited for the job.
+
+*** Part 1: Declaring constants
+First we are going to define our constants and our allocator
+#+BEGIN_SRC zig
+const allocator = std.heap.page_allocator;
+const app_name: []const u8 = "/bin/i3-msg";
+var params: []const []const u8 = &.{ app_name, "-t", "get_workspaces" };
+#+END_SRC
+
+In python we don't need the allocator since it is a garbage collected language, but we still need to define our constants and we see that the syntax is quite nicer in Python
+#+BEGIN_SRC python
+app_name = "/bin/i3-msg"
+params = app_name + " -t get_workspaces"
+#+END_SRC
+
+*** Part 2: Getting the output of the process
+Then we are going to call the process and get the output
+#+BEGIN_SRC zig
+const result = try std.ChildProcess.run(.{
+    .allocator = allocator,
+    .argv = params,
+    .cwd = null,
+    .cwd_dir = null,
+    .env_map = null,
+    .max_output_bytes = 51200,
+    .expand_arg0 = .no_expand,
+});
+defer {
+    allocator.free(result.stdout);
+    allocator.free(result.stderr);
+}
+const text_result = result.stdout;
+#+END_SRC
+The python code contains way less parameters, notably those related to memory management.
+#+BEGIN_SRC python
+result = subprocess.run(params, shell=True, capture_output=True, text=True)
+#+END_SRC
+
+*** Part 3: Making the new name
+Here we are allocating dynamic memory to the arguments and then declaring the new name of the workspace
+#+BEGIN_SRC zig
+const args = try std.process.argsAlloc(allocator);
+defer std.process.argsFree(allocator, args);
+const new_name = try std.fmt.allocPrint(allocator, "{s}: {s}", .{ args[1], args[2] });
+defer allocator.free(new_name);
+#+END_SRC
+Again, we don't have to worry about allocation and deallocating memory. Moreover, string maniuplation are way easier in Python.
+#+BEGIN_SRC python
+new_name = sys.argv[1] + ": " + sys.argv[2]
+#+END_SRC
+
+*** Part 4: Getting the current name of the workspace from the id of the workspace we want to rename
+There is no std regex functions or anything that contains string manipulations in Zig, so we have to do it by hand by analyzing the characters of the string. Obviously, this is a very error-prone and hard task.
+#+BEGIN_SRC zig
+const needle: []const u8 = try std.fmt.allocPrint(allocator, "num\":{s}", .{args[1]});
+defer allocator.free(needle);
+var pos_in_text_result = std.mem.indexOf(u8, text_result, needle);
+pos_in_text_result.? += 15;
```

Comment:
Maybe add const, because 15, 10, 3, I dont know the meaning of those numbers

Reply:
no need they are litterals